hazelcast/src/hazelcast/client/connection/ClientConnectionManagerImpl.cpp

```diff
@@ -0,0 +1,842 @@
+
+                if (!inSelector.start()) {
+                    return false;
+                }
+                if (!outSelector.start()) {
+                    return false;
+                }
+
+                startEventLoopGroup();
+                heartbeat.reset(new HeartbeatManager(client));
+                heartbeat->start();
+                addConnectionHeartbeatListener(boost::shared_ptr<spi::impl::ConnectionHeartbeatListener>(
+                        new spi::impl::ConnectionHeartbeatListenerDelegator(*this)));
+                connectionStrategy->start();
+
+                return result;
+            }
+
+            void ClientConnectionManagerImpl::shutdown() {
+                util::LockGuard guard(lock);
+                if (!alive) {
+                    return;
+                }
+                alive = false;
+
+                // close connections
+                BOOST_FOREACH(boost::shared_ptr<Connection> connection, activeConnections.values()) {
+                                // prevent any exceptions
+                                util::IOUtil::closeResource(connection.get(), "Hazelcast client is shutting down");
+                            }
+
+                spi::impl::ClientExecutionServiceImpl::shutdownExecutor("cluster", *clusterConnectionExecutor, logger);
+                stopEventLoopGroup();
+                connectionListeners.clear();
+                heartbeat->shutdown();
+
+                connectionStrategy->shutdown();
+
+
+                if (alive.compareAndSet(true, false)) {
+                    return;
+                }
+
+                activeConnectionsFileDescriptors.clear();
+                activeConnections.clear();
+                socketConnections.clear();
+            }
+
+            boost::shared_ptr<Connection>
+            ClientConnectionManagerImpl::getOrConnect(const Address &address) {
+                return getOrConnect(address, false);
+            }
+
+            boost::shared_ptr<Connection> ClientConnectionManagerImpl::getOwnerConnection() {
+                boost::shared_ptr<Address> address = ownerConnectionAddress;
+                if (address.get() == NULL) {
+                    return boost::shared_ptr<Connection>();
+                }
+                boost::shared_ptr<Connection> connection = getActiveConnection(*address);
+                return connection;
+
+            }
+
+            boost::shared_ptr<Connection> ClientConnectionManagerImpl::connectAsOwner(const Address &address) {
+                boost::shared_ptr<Connection> connection;
+                try {
+                    logger.info() << "Trying to connect to " << address << " as owner member";
+                    connection = getOrConnect(address, true);
+                    client.onClusterConnect(connection);
+                    setOwnerConnectionAddress(boost::shared_ptr<Address>(new Address(connection->getRemoteEndpoint())));
```

Comment:
The place that setOwnerConnectionAdress is set is changed due to a bug. Can you  recheck java for this and change accrodingly ? Related pr https://github.com/hazelcast/hazelcast/pull/12415/files

Reply:
Done.